File: 4kyu_Matrix_Determinant.cpp
#include <iostream>
#include <vector>
using namespace std;

long long determinant(vector< vector<long long> > m) {
	if (m.size() == 1) return m[0][0];
	vector<vector<vector<long long>>> minors;
	long long result = 0;
	
	for (int i = 0; i < (int)m.size(); i++) {
		vector<vector<long long>> minor;
		for (int y = 1; y < (int)m.size(); y++) {
			vector<long long> row;
			for (int x = 0; x < (int)m.size(); x++) {
				if (i == x) continue;
				row.push_back(m[y][x]);
			}
			minor.push_back(row);
		}
		minors.push_back(minor);
	}

		for (int i = 0; i < (int)m.size(); i++)
			result += m[0][i] * determinant(minors[i]) * ((i % 2 == 0) ? 1 : -1);

	return result;
}

int main()
{
	cout << determinant({ {1, 3}, {2, 5} });
}

File: Codewars/JavaScript/4kyu_Matrix_Determinant.js
function determinant(m) {
  const n = m.length;
  if (n === 1) return m[0][0];
  else if (n === 2) return m[0][0] * m[1][1] - m[0][1] * m[1][0];
  else {
    return m[0].reduce((prev, cur, x) => prev + cur * determinant(makeMinor(x)) * (x % 2 === 0 ? 1 : -1), 0);
  }

  function makeMinor(targetX, targetY = 0) {
    const result = [];
    for (let y = 0; y < n; y++) {
      if (y === targetY) continue;
      const row = [];
      for (let x = 0; x < n; x++) {
        if (x === targetX) continue;
        row.push(m[y][x]);
      }
      result.push(row);
    }

    return result;
  }
}

const matrix = [
  [2, 4, 2],
  [3, 1, 1],
  [1, 2, 0],
];
console.log(determinant(matrix));
